src/gpu/intel/jit/dsl/dsl.hpp

```diff
@@ -42,125 +33,22 @@ int grf_size();
-            normalized = kind_t::none;
-
-        switch (normalized) {
-            case kind_t::none:
-                return v2::layout_t(desc, type, 0,
-                        {{col_var, col, 1}, {row_var, row, col}});
-
-            case kind_t::block: {
-                int col_outer = (int)(col / col_inner);
-                return v2::layout_t(desc, type, 0,
-                        {{col_var, col_inner, 1}, {row_var, row, col_inner},
-                                {col_var, col_outer, row * col_inner}});
-            }
-
-            case kind_t::vnni: {
-                int row_inner = 4 / t;
-                int row_outer = (int)(row / row_inner);
-                int col_outer = (int)(col / col_inner);
-                return v2::layout_t(desc, type, 0,
-                        {{row_var, row_inner, 1},
-                                {col_var, col_inner, row_inner},
-                                {row_var, row_outer, col_inner * row_inner},
-                                {col_var, col_outer,
-                                        row_outer * col_inner * row_inner}});
-            }
-
-            // Impossible to hit due to normalization
-            case kind_t::transpose_vnni:
-            default: gpu_assert(false); return {};
-        }
-    }
-
-    // Tile used for 2d messages
-    tile_t get_2d_tile(type_t type) const {
-        if (kind == kind_t::transpose_vnni) {
-            auto width = pack_size ? pack_size
-                                   : grf_size() / std::max(type.size(), 4);
-            auto height = 32;
-            return {{dims[1], width}, {dims[0], height}};
-        }
-
-        auto width = pack_size ? pack_size : grf_size() / type.size();
-        auto height = 32;
-        return {{dims[0], width}, {dims[1], height}};
-    }
-
-    // Tile used for block loads
-    tile_t get_block_tile(type_t type) const {
-        if (kind == kind_t::none) {
-            return {{dims[0], 8 * grf_size() / type.size()}, {dims[1], 1}};
-        } else if (kind == kind_t::block) {
-            return {{dims[0], pack_size ? pack_size : grf_size() / type.size()},
-                    {dims[1], 1}};
-        } else {
-            gpu_assert(false);
-            return {};
-        }
-    }
-
-    static send_cache_hint_t to_ir(ngen::CacheSettingsLSC hint) {
-        switch (hint) {
-            case ngen::CacheSettingsLSC::L1C_L3C:
-                return send_cache_hint_t::load_once;
-            case ngen::CacheSettingsLSC::Default:
-                return send_cache_hint_t::hw_default;
-            default: gpu_assert(false); return send_cache_hint_t::undef;
-        }
-    }
-
-    kind_t kind = kind_t::none;
-    int pack_size = 0;
-    send_cache_hint_t cache_hint = send_cache_hint_t::undef;
-    std::array<pvar_t, 2> dims = {};
+struct send_hint_t {
+    send_cache_hint_t cache = send_cache_hint_t::undef;
 };
 
 struct tensor_t {
+    tensor_t sub_tensor(const v2::layout_t &l, const icoord_t &coord) const {
```

Comment:
A comment for the future: we may need `layout_t::map()` like interface to pass both a tile and a coordinate. Also the offset can be part of the layout to keep the buffer normalized (without offset). With that, it would be natural to align it with `layout_t::map()` both in naming and the argument list.

Reply:
I think I need to look at the implementation. To me it sounds like it's all there, if you have logical `k` offset - you can easily extract the sub-layout via `map()`, you just need the relative `k` index and the tile sizes.

At first glance, looking at the GEMM implementation here:

https://github.com/uxlfoundation/oneDNN/blob/2cbc7374fb8b13579d528cedb471291947a0a17d/src/gpu/intel/jit/gemm/ir/builder.cpp#L210-L214

I think we only need two out of these 4 objects: `copy_layout` and `tile`. Other two are derived from these.

Reply:
> I think we only need two out of these 4 objects: copy_layout and tile. Other two are derived from these.

I don't disagree, but in this case, `copy_layout` is originally derived from `layout`. While we can rederive `layout` from `copy_layout` + `tile`, it much less work to just save the already computed `layout` for use later.

Reply:
I agree, it's more computationally optimal, but it's just easier to judge on the logic if there are less entities and we derive things on the fly.

Say, here we have one storage which has a fixed layout (this layout is one entity). Now to the loop: here we have indices and tile sizes corresponding to the algorithm. With these three entities we can construct loads and multiplication.

If 1) we can avoid saving derived temporary objects and 2) it's not significantly slower, IMO, better to use a more compact style.